Crash: java.lang.IllegalStateException: Not allowed to start service Intent

Crash collected on : Thu Aug 11 11:05:48 GMT+02:00 2022

Informations :
Locale: en_US
Version: 1.3.0-cont-integ_94ec3cf5_220806-1540 CI
Package: [com.poupa.vinylmusicplayer.ci](http://com.poupa.vinylmusicplayer.ci/)
Phone Model: ONEPLUS A5000
Android Version: 11
Board: msm8998
Brand: OnePlus
Device: OnePlus5
Host: 8d79c3c0af02
ID: RQ3A.211001.001
Model: ONEPLUS A5000
Product: OnePlus5
Type: userdebug
Total Internal memory: 119115218944
Available Internal memory: 21468790784


Stack:
```
java.lang.RuntimeException: Unable to start activity ComponentInfo{[com.poupa.vinylmusicplayer.ci/com.poupa.vinylmusicplayer.ui.activities.MainActivity](http://com.poupa.vinylmusicplayer.ci/com.poupa.vinylmusicplayer.ui.activities.MainActivity)}: java.lang.IllegalStateException: Not allowed to start service Intent { cmp=[com.poupa.vinylmusicplayer.ci/com.poupa.vinylmusicplayer.service.MusicService](http://com.poupa.vinylmusicplayer.ci/com.poupa.vinylmusicplayer.service.MusicService) }: app is in background uid UidRecord{b09b4d7 u0a206 RCVR idle change:uncached procs:1 seq(0,0,0)}
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3431)
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3595)
at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:223)
at android.app.ActivityThread.main(ActivityThread.java:7664)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
Caused by: java.lang.IllegalStateException: Not allowed to start service Intent { cmp=[com.poupa.vinylmusicplayer.ci/com.poupa.vinylmusicplayer.service.MusicService](http://com.poupa.vinylmusicplayer.ci/com.poupa.vinylmusicplayer.service.MusicService) }: app is in background uid UidRecord{b09b4d7 u0a206 RCVR idle change:uncached procs:1 seq(0,0,0)}
at android.app.ContextImpl.startServiceCommon(ContextImpl.java:1715)
at android.app.ContextImpl.startService(ContextImpl.java:1670)
at android.content.ContextWrapper.startService(ContextWrapper.java:720)
at android.content.ContextWrapper.startService(ContextWrapper.java:720)
at android.content.ContextWrapper.startService(ContextWrapper.java:720)
at com.poupa.vinylmusicplayer.helper.MusicPlayerRemote.bindToService(Unknown Source:22)
at com.poupa.vinylmusicplayer.ui.activities.base.AbsMusicServiceActivity.onCreate(Unknown Source:8)
at com.poupa.vinylmusicplayer.ui.activities.base.AbsSlidingMusicPanelActivity.onCreate(Unknown Source:0)
at com.poupa.vinylmusicplayer.ui.activities.MainActivity.onCreate(Unknown Source:0)
at android.app.Activity.performCreate(Activity.java:7994)
at android.app.Activity.performCreate(Activity.java:7978)
at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3404)
... 11 more
```

--- a/app/src/main/java/com/poupa/vinylmusicplayer/appshortcuts/AppShortcutLauncherActivity.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/appshortcuts/AppShortcutLauncherActivity.java
@@ -4,6 +4,8 @@ import android.app.Activity;
 import android.content.Intent;
 import android.os.Bundle;
 
+import androidx.core.content.ContextCompat;
+
 import com.poupa.vinylmusicplayer.appshortcuts.shortcuttype.LastAddedShortcutType;
 import com.poupa.vinylmusicplayer.appshortcuts.shortcuttype.ShuffleAllShortcutType;
 import com.poupa.vinylmusicplayer.appshortcuts.shortcuttype.TopTracksShortcutType;
@@ -69,6 +71,6 @@ public class AppShortcutLauncherActivity extends Activity {
 
         intent.putExtras(bundle);
 
-        startService(intent);
+        ContextCompat.startForegroundService(this, intent);
     }
 }

--- a/app/src/main/java/com/poupa/vinylmusicplayer/helper/MusicPlayerRemote.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/helper/MusicPlayerRemote.java
@@ -20,6 +20,7 @@ import android.widget.Toast;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
+import androidx.core.content.ContextCompat;
 
 import com.poupa.vinylmusicplayer.R;
 import com.poupa.vinylmusicplayer.discog.Discography;
@@ -53,14 +54,13 @@ public class MusicPlayerRemote {
             realActivity = (Activity) context;
         }
 
-        final ContextWrapper contextWrapper = new ContextWrapper(realActivity);
-        contextWrapper.startService(new Intent(contextWrapper, MusicService.class));
+        ContextCompat.startForegroundService(realActivity, new Intent(realActivity, MusicService.class));
 
         final ServiceBinder binder = new ServiceBinder(callback);
 
-        if (contextWrapper.bindService(new Intent().setClass(contextWrapper, MusicService.class), binder, Context.BIND_AUTO_CREATE)) {
-            mConnectionMap.put(contextWrapper, binder);
-            return new ServiceToken(contextWrapper);
+        if (realActivity.bindService(new Intent().setClass(realActivity, MusicService.class), binder, Context.BIND_AUTO_CREATE)) {
+            mConnectionMap.put(realActivity, binder);
+            return new ServiceToken(realActivity);
         }
         return null;
     }

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/MediaButtonIntentReceiver.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/MediaButtonIntentReceiver.java
@@ -169,21 +169,7 @@ public class MediaButtonIntentReceiver extends BroadcastReceiver {
     private static void startService(Context context, String command) {
         final Intent intent = new Intent(context, MusicService.class);
         intent.setAction(command);
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
-            context.startForegroundService(intent);
-        } else {
-            try {
-                // IMPORTANT NOTE: (kind of a hack)
-                // on Android O and above the following crashes when the app is not running
-                // there is no good way to check whether the app is running so we catch the exception
-                // we do not always want to use startForegroundService() because then one gets an ANR
-                // if no notification is displayed via startForeground()
-                // according to Play analytics this happens a lot, I suppose for example if command = PAUSE
-                context.startService(intent);
-            } catch (IllegalStateException ignored) {
-                ContextCompat.startForegroundService(context, intent);
-            }
-        }
+        ContextCompat.startForegroundService(context, intent);
     }
 
     private static void acquireWakeLockAndSendMessage(Context context, Message msg, long delay) {

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/MusicService.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/MusicService.java
@@ -58,6 +58,7 @@ import com.poupa.vinylmusicplayer.model.Song;
 import com.poupa.vinylmusicplayer.provider.HistoryStore;
 import com.poupa.vinylmusicplayer.provider.MusicPlaybackQueueStore;
 import com.poupa.vinylmusicplayer.provider.SongPlayCountStore;
+import com.poupa.vinylmusicplayer.service.notification.IdleNotification;
 import com.poupa.vinylmusicplayer.service.notification.PlayingNotification;
 import com.poupa.vinylmusicplayer.service.notification.PlayingNotificationImpl;
 import com.poupa.vinylmusicplayer.service.notification.PlayingNotificationImpl24;
@@ -153,7 +154,10 @@ public class MusicService extends MediaBrowserServiceCompat implements SharedPre
 
     private boolean queuesRestored;
     private boolean pausedByTransientLossOfFocus;
+
     private PlayingNotification playingNotification;
+    private IdleNotification idleNotification;
+
     private AudioManager audioManager;
     private MediaSessionCompat mediaSession;
     private PowerManager.WakeLock wakeLock;
@@ -423,7 +427,10 @@ public class MusicService extends MediaBrowserServiceCompat implements SharedPre
 
     public void quit() {
         pause();
+
         playingNotification.stop();
+        idleNotification.stop();
+        stopForeground(true);
 
         closeAudioEffectSession();
         getAudioManager().abandonAudioFocus(audioFocusListener);
@@ -525,19 +532,32 @@ public class MusicService extends MediaBrowserServiceCompat implements SharedPre
         return (getAudioManager().requestAudioFocus(audioFocusListener, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN) == AudioManager.AUDIOFOCUS_REQUEST_GRANTED);
     }
 
-    public void initNotification() {
+    private void initNotification() {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && !PreferenceUtil.getInstance().classicNotification()) {
             playingNotification = new PlayingNotificationImpl24();
         } else {
             playingNotification = new PlayingNotificationImpl();
         }
-        playingNotification.init(this);
+        playingNotification.init(this,
+                PlayingNotification.NOTIFICATION_CHANNEL_ID,
+                R.string.playing_notification_name,
+                R.string.playing_notification_description);
+
+        idleNotification = new IdleNotification();
+        idleNotification.init(this,
+                IdleNotification.NOTIFICATION_CHANNEL_ID,
+                R.string.idle_notification_name,
+                R.string.idle_notification_description);
     }
 
-    public void updateNotification() {
-        //if (playingNotification != null && getCurrentSong().id != Song.EMPTY_SONG.id) {
-        if (playingNotification != null) {
+    private void updateNotification() {
+        if (getCurrentSong().id != Song.EMPTY_SONG.id) {
+            idleNotification.stop();
             playingNotification.update();
+        }
+        else {
+            playingNotification.stop();
+            idleNotification.update();
         }
     }
 
@@ -1027,7 +1047,7 @@ public class MusicService extends MediaBrowserServiceCompat implements SharedPre
                 if (playingQueue.size() > 0) {
                     prepareNext();
                 } else {
-                    playingNotification.stop();
+                    updateNotification();
                 }
                 break;
         }

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/MusicService.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/MusicService.java
@@ -221,7 +221,10 @@ public class MusicService extends MediaBrowserServiceCompat implements SharedPre
         registerReceiver(updateFavoriteReceiver, new IntentFilter(FAVORITE_STATE_CHANGED));
 
         initNotification();
-
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            // Android O+ requires a foreground service to post a notification asap
+            updateNotification();
+        }
         mediaStoreObserver = new MediaStoreObserver(this, playerHandler);
         throttledSeekHandler = new ThrottledSeekHandler(this, playerHandler);
         getContentResolver().registerContentObserver(
@@ -532,7 +535,8 @@ public class MusicService extends MediaBrowserServiceCompat implements SharedPre
     }
 
     public void updateNotification() {
-        if (playingNotification != null && getCurrentSong().id != -1) {
+        //if (playingNotification != null && getCurrentSong().id != Song.EMPTY_SONG.id) {
+        if (playingNotification != null) {
             playingNotification.update();
         }
     }

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/AbsNotification.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/AbsNotification.java
@@ -1,0 +1,49 @@
+package com.poupa.vinylmusicplayer.service.notification;
+
+import android.app.NotificationChannel;
+import android.app.NotificationManager;
+import android.content.Context;
+import android.os.Build;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
+import androidx.annotation.StringRes;
+
+import com.poupa.vinylmusicplayer.service.MusicService;
+
+public abstract class AbsNotification {
+    NotificationManager notificationManager;
+    protected MusicService service;
+
+    public void init(final MusicService musicService, @NonNull final String channelId,
+                     @StringRes final int channelName,
+                     @StringRes final int channelDescription) {
+        service = musicService;
+        notificationManager = (NotificationManager) service.getSystemService(Context.NOTIFICATION_SERVICE);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            createNotificationChannel(channelId, channelName, channelDescription);
+        }
+    }
+
+    @RequiresApi(26)
+    private void createNotificationChannel(@NonNull final String channelId, @StringRes final int channelName,
+                                           @StringRes final int channelDescription) {
+        NotificationChannel notificationChannel = notificationManager.getNotificationChannel(channelId);
+        if (notificationChannel == null) {
+            notificationChannel = new NotificationChannel(
+                    channelId,
+                    service.getString(channelName),
+                    NotificationManager.IMPORTANCE_LOW);
+            notificationChannel.setDescription(service.getString(channelDescription));
+            notificationChannel.enableLights(false);
+            notificationChannel.enableVibration(false);
+
+            notificationManager.createNotificationChannel(notificationChannel);
+        }
+    }
+
+    public abstract void update();
+
+    public abstract void stop();
+
+}

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/IdleNotification.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/IdleNotification.java
@@ -14,7 +14,7 @@ public class IdleNotification extends AbsNotification {
         final Notification notification = new NotificationCompat.Builder(service, NOTIFICATION_CHANNEL_ID)
                 .setSmallIcon(R.drawable.ic_notification)
                 .setContentTitle(service.getString(R.string.idle_notification_title))
-                .setContentText(service.getString(R.string.idle_notification_description))
+                .setOngoing(true)
                 .setCategory(NotificationCompat.CATEGORY_SERVICE)
                 .setPriority(NotificationCompat.PRIORITY_LOW)
                 .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/IdleNotification.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/IdleNotification.java
@@ -1,0 +1,28 @@
+package com.poupa.vinylmusicplayer.service.notification;
+
+import android.app.Notification;
+
+import androidx.core.app.NotificationCompat;
+
+import com.poupa.vinylmusicplayer.R;
+
+public class IdleNotification extends AbsNotification {
+    private static final int NOTIFICATION_ID = 2;
+    public static final String NOTIFICATION_CHANNEL_ID = "idle_notification";
+
+    public void update() {
+        final Notification notification = new NotificationCompat.Builder(service, NOTIFICATION_CHANNEL_ID)
+                .setSmallIcon(R.drawable.ic_notification)
+                .setContentTitle(service.getString(R.string.idle_notification_title))
+                .setContentText(service.getString(R.string.idle_notification_description))
+                .setCategory(NotificationCompat.CATEGORY_SERVICE)
+                .setPriority(NotificationCompat.PRIORITY_LOW)
+                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
+                .build();
+        service.startForeground(NOTIFICATION_ID, notification);
+    }
+
+    public void stop() {
+        notificationManager.cancel(NOTIFICATION_ID);
+    }
+}

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotification.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotification.java
@@ -1,80 +1,21 @@
 package com.poupa.vinylmusicplayer.service.notification;
 
 import android.app.Notification;
-import android.app.NotificationChannel;
-import android.app.NotificationManager;
-import android.os.Build;
 
-import androidx.annotation.RequiresApi;
-
-import com.poupa.vinylmusicplayer.R;
-import com.poupa.vinylmusicplayer.service.MusicService;
-
-import static android.content.Context.NOTIFICATION_SERVICE;
-
-public abstract class PlayingNotification {
-
+public abstract class PlayingNotification extends AbsNotification {
     private static final int NOTIFICATION_ID = 1;
-    static final String NOTIFICATION_CHANNEL_ID = "playing_notification";
+    public static final String NOTIFICATION_CHANNEL_ID = "playing_notification";
 
-    //private static final int NOTIFY_MODE_FOREGROUND = 1;
-    //private static final int NOTIFY_MODE_BACKGROUND = 0;
-
-    //private int notifyMode = NOTIFY_MODE_BACKGROUND;
-
-    private NotificationManager notificationManager;
-    protected MusicService service;
     boolean stopped;
-
-    public synchronized void init(MusicService service) {
-        this.service = service;
-        notificationManager = (NotificationManager) service.getSystemService(NOTIFICATION_SERVICE);
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
-            createNotificationChannel();
-        }
-    }
 
     public abstract void update();
 
-    public synchronized void stop() {
+    public void stop() {
         stopped = true;
-        service.stopForeground(true);
         notificationManager.cancel(NOTIFICATION_ID);
     }
 
-    void updateNotifyModeAndPostNotification(Notification notification) {
+    void updateImpl(Notification notification) {
         service.startForeground(NOTIFICATION_ID, notification);
-
-        //int newNotifyMode;
-        //if (service.isPlaying()) {
-        //    newNotifyMode = NOTIFY_MODE_FOREGROUND;
-        //} else {
-        //    newNotifyMode = NOTIFY_MODE_BACKGROUND;
-        //}
-        //
-        //if (notifyMode != newNotifyMode && newNotifyMode == NOTIFY_MODE_BACKGROUND) {
-        //    service.stopForeground(false);
-        //}
-        //
-        //if (newNotifyMode == NOTIFY_MODE_FOREGROUND) {
-        //    service.startForeground(NOTIFICATION_ID, notification);
-        //} else if (newNotifyMode == NOTIFY_MODE_BACKGROUND) {
-        //    notificationManager.notify(NOTIFICATION_ID, notification);
-        //}
-        //
-        //notifyMode = newNotifyMode;
-    }
-
-    @RequiresApi(26)
-    private void createNotificationChannel() {
-        NotificationChannel notificationChannel = notificationManager.getNotificationChannel(NOTIFICATION_CHANNEL_ID);
-        if (notificationChannel == null) {
-            notificationChannel = new NotificationChannel(NOTIFICATION_CHANNEL_ID, service.getString(R.string.playing_notification_name), NotificationManager.IMPORTANCE_LOW);
-            notificationChannel.setDescription(service.getString(R.string.playing_notification_description));
-            notificationChannel.enableLights(false);
-            notificationChannel.enableVibration(false);
-
-            notificationManager.createNotificationChannel(notificationChannel);
-        }
     }
 }

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotification.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotification.java
@@ -17,10 +17,10 @@ public abstract class PlayingNotification {
     private static final int NOTIFICATION_ID = 1;
     static final String NOTIFICATION_CHANNEL_ID = "playing_notification";
 
-    private static final int NOTIFY_MODE_FOREGROUND = 1;
-    private static final int NOTIFY_MODE_BACKGROUND = 0;
+    //private static final int NOTIFY_MODE_FOREGROUND = 1;
+    //private static final int NOTIFY_MODE_BACKGROUND = 0;
 
-    private int notifyMode = NOTIFY_MODE_BACKGROUND;
+    //private int notifyMode = NOTIFY_MODE_BACKGROUND;
 
     private NotificationManager notificationManager;
     protected MusicService service;
@@ -43,24 +43,26 @@ public abstract class PlayingNotification {
     }
 
     void updateNotifyModeAndPostNotification(Notification notification) {
-        int newNotifyMode;
-        if (service.isPlaying()) {
-            newNotifyMode = NOTIFY_MODE_FOREGROUND;
-        } else {
-            newNotifyMode = NOTIFY_MODE_BACKGROUND;
-        }
+        service.startForeground(NOTIFICATION_ID, notification);
 
-        if (notifyMode != newNotifyMode && newNotifyMode == NOTIFY_MODE_BACKGROUND) {
-            service.stopForeground(false);
-        }
-
-        if (newNotifyMode == NOTIFY_MODE_FOREGROUND) {
-            service.startForeground(NOTIFICATION_ID, notification);
-        } else if (newNotifyMode == NOTIFY_MODE_BACKGROUND) {
-            notificationManager.notify(NOTIFICATION_ID, notification);
-        }
-
-        notifyMode = newNotifyMode;
+        //int newNotifyMode;
+        //if (service.isPlaying()) {
+        //    newNotifyMode = NOTIFY_MODE_FOREGROUND;
+        //} else {
+        //    newNotifyMode = NOTIFY_MODE_BACKGROUND;
+        //}
+        //
+        //if (notifyMode != newNotifyMode && newNotifyMode == NOTIFY_MODE_BACKGROUND) {
+        //    service.stopForeground(false);
+        //}
+        //
+        //if (newNotifyMode == NOTIFY_MODE_FOREGROUND) {
+        //    service.startForeground(NOTIFICATION_ID, notification);
+        //} else if (newNotifyMode == NOTIFY_MODE_BACKGROUND) {
+        //    notificationManager.notify(NOTIFICATION_ID, notification);
+        //}
+        //
+        //notifyMode = newNotifyMode;
     }
 
     @RequiresApi(26)

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotification.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotification.java
@@ -10,7 +10,7 @@ public abstract class PlayingNotification extends AbsNotification {
 
     public abstract void update();
 
-    public void stop() {
+    public synchronized void stop() {
         stopped = true;
         notificationManager.cancel(NOTIFICATION_ID);
     }

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl.java
@@ -40,7 +40,7 @@ public class PlayingNotificationImpl extends PlayingNotification {
     private Target target;
 
     @Override
-    public synchronized void update() {
+    public void update() {
         stopped = false;
 
         final Song song = service.getCurrentSong();
@@ -130,7 +130,7 @@ public class PlayingNotificationImpl extends PlayingNotification {
 
                                 if (stopped)
                                     return; // notification has been stopped before loading was finished
-                                updateNotifyModeAndPostNotification(notification);
+                                updateImpl(notification);
                             }
 
                             private void setBackgroundColor(int color) {

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl.java
@@ -40,7 +40,7 @@ public class PlayingNotificationImpl extends PlayingNotification {
     private Target target;
 
     @Override
-    public void update() {
+    public synchronized void update() {
         stopped = false;
 
         final Song song = service.getCurrentSong();

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl24.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl24.java
@@ -36,7 +36,7 @@ import static com.poupa.vinylmusicplayer.service.MusicService.TOGGLE_FAVORITE;
 public class PlayingNotificationImpl24 extends PlayingNotification {
 
     @Override
-    public void update() {
+    public synchronized void update() {
         stopped = false;
 
         final Song song = service.getCurrentSong();

--- a/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl24.java
+++ b/app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl24.java
@@ -36,7 +36,7 @@ import static com.poupa.vinylmusicplayer.service.MusicService.TOGGLE_FAVORITE;
 public class PlayingNotificationImpl24 extends PlayingNotification {
 
     @Override
-    public synchronized void update() {
+    public void update() {
         stopped = false;
 
         final Song song = service.getCurrentSong();
@@ -114,7 +114,7 @@ public class PlayingNotificationImpl24 extends PlayingNotification {
 
                         if (stopped)
                             return; // notification has been stopped before loading was finished
-                        updateNotifyModeAndPostNotification(builder.build());
+                        updateImpl(builder.build());
                     }
                 }));
     }

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -294,8 +294,14 @@
     <string name="playlist_is_empty">Playlist is empty</string>
     <string name="snack_bar_action_undo">undo</string>
     <string name="snack_bar_title_removed_song">Removed</string>
+
     <string name="playing_notification_description">The playing notification provides actions for play/pause etc.</string>
     <string name="playing_notification_name">Playing notification</string>
+
+    <string name="idle_notification_description">The application is ready but the playing queue is empty</string>
+    <string name="idle_notification_name">Idle notification</string>
+    <string name="idle_notification_title">Vinyl is running</string>
+
     <string name="eugene_cheung_summary">For his contributions to the source code.</string>
     <string name="adrian_summary">For creating the album page design.</string>
     <string name="add_action">Add</string>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -298,7 +298,7 @@
     <string name="playing_notification_description">The playing notification provides actions for play/pause etc.</string>
     <string name="playing_notification_name">Playing notification</string>
 
-    <string name="idle_notification_description">The application is ready but the playing queue is empty</string>
+    <string name="idle_notification_description">The idle notification is used where the playing queue is empty</string>
     <string name="idle_notification_name">Idle notification</string>
     <string name="idle_notification_title">Vinyl is running</string>
 

